pkg/k8s/watchers/cilium_node.go

```diff
@@ -5,153 +5,159 @@ package watchers
+					return
+				case event, ok := <-events:
+					if !ok {
+						return
+					}
+					var err error
+					switch event.Kind {
+					case resource.Sync:
+						synced.Store(true)
+					case resource.Upsert:
+						var needUpdate bool
+						oldObj, ok := cache[event.Key]
+						if !ok {
+							needUpdate = k.onCiliumNodeInsert(event.Object)
+						} else {
+							needUpdate = k.onCiliumNodeUpdate(oldObj, event.Object)
+						}
+						if needUpdate {
+							cache[event.Key] = event.Object.DeepCopy()
+						}
+					case resource.Delete:
+						k.onCiliumNodeDelete(event.Object)
+						delete(cache, event.Key)
+					}
+					event.Done(err)
+				}
+			}
+		}()
 
-		close(isConnected)
+		select {
+		case <-ctx.Done():
+			return
+		case <-kvstore.Client().Disconnected():
+			log.Info("Disconnected from key-value store, restarting CiliumNode watcher")
+		}
+	}
+}
 
-		k.cancelWaitGroupToSyncResources(apiGroup)
-		k.k8sAPIGroups.RemoveAPI(apiGroup)
-		// Create a new node controller when we are disconnected with the
-		// kvstore
-		<-kvstore.Client().Disconnected()
+func (k *K8sWatcher) onCiliumNodeInsert(ciliumNode *cilium_v2.CiliumNode) bool {
+	if k8s.IsLocalCiliumNode(ciliumNode) {
+		return false
+	}
+	n := types.ParseCiliumNode(ciliumNode)
+	k.nodeDiscoverManager.NodeUpdated(n)
+	return true
+}
 
-		log.Info("Disconnected from key-value store, restarting CiliumNode watcher")
+func (k *K8sWatcher) onCiliumNodeUpdate(oldNode, newNode *cilium_v2.CiliumNode) bool {
+	// Comparing Annotations here since wg-pub-key annotation is used to exchange rotated WireGuard keys.
+	if oldNode.DeepEqual(newNode) &&
+		comparator.MapStringEquals(oldNode.ObjectMeta.Labels, newNode.ObjectMeta.Labels) &&
+		comparator.MapStringEquals(oldNode.ObjectMeta.Annotations, newNode.ObjectMeta.Annotations) {
+		return false
 	}
+	return k.onCiliumNodeInsert(newNode)
+}
+
+func (k *K8sWatcher) onCiliumNodeDelete(ciliumNode *cilium_v2.CiliumNode) {
+	n := types.ParseCiliumNode(ciliumNode)
+	k.nodeDiscoverManager.NodeDeleted(n)
 }
 
-// GetCiliumNode returns the CiliumNode "nodeName" from the local store. If the
-// local store is not initialized then it will fallback retrieving the node
+// GetCiliumNode returns the CiliumNode "nodeName" from the local Resource[T] store. If the
+// local Resource[T] store is not initialized then it will fallback retrieving the node
 // from kube-apiserver.
 func (k *K8sWatcher) GetCiliumNode(ctx context.Context, nodeName string) (*cilium_v2.CiliumNode, error) {
-	var (
-		err           error
-		nodeInterface interface{}
-		exists        bool
-	)
-	k.ciliumNodeStoreMU.RLock()
-	// k.ciliumNodeStore might not be set in all invocations of GetCiliumNode,
-	// for example, when the key-value store is connected and the local store might
-	// have stale versions of the CiliumNode resource.
-	if k.ciliumNodeStore == nil {
+	// In case this function is called before starting the Events stream from ciliumNodeInit,
```

Comment:
I find this a bit confusing, since `forceCiliumNodeFromApiServer` below doesn't actually use the Resource store. Maybe it makes sense under which circumstances the bool is true?

Reply:
Yeah LGTM. I think I'd move the `if the key-value store is connected ...` just before `if k.forceCiliumNodeFromApiServer`

Reply:
Done!